### PR TITLE
Setup infrastructure to port the middle-end passes to the NPM

### PR DESCRIPTION
**Pull request by bsaleil**
This patch is bigger but very similar to the one I posted to setup the infrastructure for the frontend passes. This is an NFC patch, the new pass manager is disabled by default for the middle-end passes and will still be disabled by default after #1419 is merged.

### CODE REVIEW

**Review by piotrAMD — DISMISSED** [review]

**Review by amdrexu — DISMISSED** [review]

**Comment by bsaleil** [comment]
Fixed merge conflict

**Review by kuhar — COMMENTED** [review]
FYI, clang-tidy reports naming convention violations: https://github.com/GPUOpen-Drivers/llpc/runs/4050988893?check_suite_focus=true

**Comment by bsaleil** [comment]
Yep, I'll fix that.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a7f95853726c4d4e27fffd15db4e4c45cdd30051
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev ba44ed04a5abead4d83a92bd58b3704fb5890abd</li>
<li>SPVGEN: dev df321452490fd5f9cf486ae3b52dbb14caff0dcf</li>
<li>XGL: dev 9bc7ee67f45707ca85c375bc793f7188af9edc12</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 16f6bc2bf6ccf3109a935bd9d5e43c2d98dfe8bd</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by bsaleil** [comment]
Fixed relevant clang-tidy warnings.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 76c80a03f4e7a01fb3b1ca52f0e0ff3727f33c85
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev ba44ed04a5abead4d83a92bd58b3704fb5890abd</li>
<li>SPVGEN: dev df321452490fd5f9cf486ae3b52dbb14caff0dcf</li>
<li>XGL: dev 9bc7ee67f45707ca85c375bc793f7188af9edc12</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 16f6bc2bf6ccf3109a935bd9d5e43c2d98dfe8bd</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by bsaleil** [comment]
The remaining naming convention violation (`static llvm::AnalysisKey Key;`) cannot be fixed because this variable name is required by LLVM.

File: llpc/lower/llpcSpirvLowerResourceCollect.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@
 #include <algorithm>
 // TODO: Fix the code in this file so it does not break the builder abstraction. It should
 // not be including files directly in the LGC directory tree like this.
-#include "../../lgc/builder/BuilderRecorder.h"
+#include "../../lgc/include/lgc/builder/BuilderRecorder.h"
```
Comment (kuhar) [inline llpc/lower/llpcSpirvLowerResourceCollect.cpp:39]: Can't we replace this with something that avoids the leading `../`?

File: lgc/patch/PatchNullFragShader.cpp [hunk 2]
```diff
@@ -48,43 +49,64 @@ namespace lgc {
...
-ModulePass *lgc::createPatchNullFragShader() {
-  return new PatchNullFragShader();
+ModulePass *lgc::createLegacyPatchNullFragShader() {
+  return new LegacyPatchNullFragShader();
+}
+
+// =====================================================================================================================
+// Run the pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
```
Comment (kuhar) [inline lgc/patch/PatchNullFragShader.cpp:83]: missing `@returns`

File: lgc/patch/PatchNullFragShader.cpp [hunk 3]
```diff
@@ -48,43 +49,64 @@ namespace lgc {
...
+  return m_impl.runImpl(module, pipelineState);
 }
 
 // =====================================================================================================================
 // Run the pass on the specified LLVM module.
 //
 // @param [in/out] module : LLVM module to be run on
-bool PatchNullFragShader::runOnModule(Module &module) {
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses PatchNullFragShader::run(Module &module, ModuleAnalysisManager &analysisManager) {
```
Comment (kuhar) [inline lgc/patch/PatchNullFragShader.cpp:94]: missing `@returns` and `@params module`

File: lgc/patch/PatchNullFragShader.cpp [hunk 4]
```diff
@@ -48,43 +49,64 @@ namespace lgc {
...
+  PipelineState *pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  if (runImpl(module, pipelineState))
+    return PreservedAnalyses::none();
+  return PreservedAnalyses::all();
+}
+
+// =====================================================================================================================
+// Run the pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
```
Comment (kuhar) [inline lgc/patch/PatchNullFragShader.cpp:104]: missing `@returns` and `@params pipelineState`

File: lgc/state/PipelineState.cpp [hunk 5]
```diff
@@ -1499,61 +1499,118 @@ PipelineState *PipelineStateWrapper::getPipelineState(Module *module) {
   return m_pipelineState;
 }
 
+// =====================================================================================================================
+// Execute this analysis on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this analysis
+PipelineStateWrapper::Result PipelineStateWrapper::run(Module &module, ModuleAnalysisManager &analysisManager) {
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1507]: missing `@returns`

File: lgc/state/PipelineState.cpp [hunk 6]
```diff
@@ -1499,61 +1499,118 @@ PipelineState *PipelineStateWrapper::getPipelineState(Module *module) {
...
-ModulePass *lgc::createPipelineStateClearer() {
-  return new PipelineStateClearer();
+ModulePass *lgc::createLegacyPipelineStateClearer() {
+  return new LegacyPipelineStateClearer();
+}
+
+// =====================================================================================================================
+// Run PipelineStateClearer pass to clear the pipeline state out of the IR
+//
+// @param [in/out] module : IR module
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1545]: missing `@returns`

File: lgc/state/PipelineState.cpp [hunk 7]
```diff
@@ -1499,61 +1499,118 @@ PipelineState *PipelineStateWrapper::getPipelineState(Module *module) {
...
+  PipelineState *pipelineState = getAnalysis<LegacyPipelineStateWrapper>().getPipelineState(&module);
+  return m_impl.runImpl(module, pipelineState);
+}
+
+// =====================================================================================================================
+// Run PipelineStateClearer pass to clear the pipeline state out of the IR
+//
+// @param [in/out] module : IR module
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses PipelineStateClearer::run(Module &module, ModuleAnalysisManager &analysisManager) {
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1556]: missing `@returns`

File: lgc/state/PipelineState.cpp [hunk 8]
```diff
@@ -1499,61 +1499,118 @@ PipelineState *PipelineStateWrapper::getPipelineState(Module *module) {
...
+  return PreservedAnalyses::none();
 }
 
 // =====================================================================================================================
 // Run PipelineStateClearer pass to clear the pipeline state out of the IR
 //
 // @param [in/out] module : IR module
-bool PipelineStateClearer::runOnModule(Module &module) {
-  auto pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+bool PipelineStateClearer::runImpl(Module &module, PipelineState *pipelineState) {
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1566]: missing `@returns`

File: lgc/state/PipelineState.cpp [hunk 9]
```diff
@@ -1499,61 +1499,118 @@ PipelineState *PipelineStateWrapper::getPipelineState(Module *module) {
...
     : ImmutablePass(ID), m_builderContext(builderContext) {
 }
 
 // =====================================================================================================================
-// Clean-up of PipelineStateWrapper at end of pass manager run
+// Clean-up of LegacyPipelineStateWrapper at end of pass manager run
 //
 // @param module : Module
-bool PipelineStateWrapper::doFinalization(Module &module) {
+bool LegacyPipelineStateWrapper::doFinalization(Module &module) {
```
Comment (kuhar) [inline lgc/state/PipelineState.cpp:1610]: missing `@returns`

File: lgc/builder/BuilderReplayer.cpp [hunk 10]
```diff
@@ -43,63 +44,81 @@ namespace {
...
+//
+// @param pipeline : Pipeline object
+LegacyBuilderReplayer::LegacyBuilderReplayer(Pipeline *pipeline) : ModulePass(ID), m_impl(pipeline) {
+}
+
+// =====================================================================================================================
+// Run the BuilderReplayer pass on a module
+//
+// @param [in/out] module : LLVM module to be run on
+bool LegacyBuilderReplayer::runOnModule(Module &module) {
```
Comment (kuhar) [inline lgc/builder/BuilderReplayer.cpp:98]: missing `@returns`

File: lgc/builder/BuilderReplayer.cpp [hunk 11]
```diff
@@ -43,63 +44,81 @@ namespace {
...
+  PipelineState *pipelineState = getAnalysis<LegacyPipelineStateWrapper>().getPipelineState(&module);
+  return m_impl.runImpl(module, pipelineState);
+}
+
+// =====================================================================================================================
+// Run the BuilderReplayer pass on a module
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+PreservedAnalyses BuilderReplayer::run(Module &module, ModuleAnalysisManager &analysisManager) {
```
Comment (kuhar) [inline lgc/builder/BuilderReplayer.cpp:108]: missing `@returns`

**Review by kuhar — COMMENTED** [review]
I found some missing documentation.

File: llpc/lower/llpcSpirvLowerResourceCollect.cpp (hunk 1 above)
Comment (bsaleil) [reply llpc/lower/llpcSpirvLowerResourceCollect.cpp:39]: As far as I understand, not without modifying the code of the pass, which is not in the scope of this patch.

**Comment by bsaleil** [comment]
Added missing documentation

**Comment by amdvlk-admin** [comment]
### Test summary for commit 57279f0bcb3e23c12566c5405c8abaae4350a68c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev ba44ed04a5abead4d83a92bd58b3704fb5890abd</li>
<li>SPVGEN: dev df321452490fd5f9cf486ae3b52dbb14caff0dcf</li>
<li>XGL: dev 9bc7ee67f45707ca85c375bc793f7188af9edc12</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 73b4a76acc8bfac6be5b551558f948b2a1492830</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@bsaleil please fix clang-tidy error:
https://github.com/GPUOpen-Drivers/llpc/runs/4070028357?check_suite_focus=true

**Comment by bsaleil** [comment]
@JaxLinAMD I cannot fix them. `clang-tidy` incorrectly reports a missing header guard and unused functions. I also cannot fix the invalid case for the `Key` variable because this specific name is required by LLVM.

**Comment by kuhar** [comment]
> @JaxLinAMD I cannot fix them. `clang-tidy` incorrectly reports a missing header guard and unused functions. I also cannot fix the invalid case for the `Key` variable because this specific name is required by LLVM.

For `Key`, you can add `// NOLINT` to silence these warnings.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ac88439d7620c8be9acabc0a5c1d9be3959b4e8f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 6cc0a035259663610287d8d99262505e78e01db2</li>
<li>SPVGEN: dev efa3cec20c6ebc66254ff9cca9b3ea124811d7a3</li>
<li>XGL: dev 0371b9a44f2dc8b1578dc2e1281b6dea439602c9</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 73b4a76acc8bfac6be5b551558f948b2a1492830</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225388)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31425/56347 (55.8%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24922/56347 (44.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31413/56347 (55.7%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24934/56347 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31748/56347 (56.3%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 24599/56347 (43.7%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34582/56347 (61.4%)</li>
<li>Failed: 0/56347 (0.0%)</li>
<li>Not Supported: 21765/56347 (38.6%)</li>
</ul></details>


</ul></details>

**Comment by bsaleil** [comment]
Added `// NOLINT` to silence `Key` variable warnings.

**Review by kuhar — APPROVED** [review]

**Review by amdrexu — APPROVED** [review]
LGTM

**Pull request merged**